### ISSUE TITLE
Composite key values in a many to many relationship are always generated on calls to AddAsync

### ISSUE THREAD

**Issue by thomasrea0113** [dotnet/efcore#19796]
I have the following models:

```csharp
    public class UserProfileModel : IDbActive, IDbEntity<UserProfileModel>
    {
        [Key]
        public string UserId { get; set; }

        // This can't be lazy loaded, because the IdentityUser model doesn't have a Profile property
        public virtual IdentityUser User { get; set; }

        public virtual ICollection<UserLeaderboard> UserLeaderboards { get; set; } = new List<UserLeaderboard>();

        public bool? IsActive { get; set; }

        public void OnModelCreating(EntityTypeBuilder<UserProfileModel> builder)
        {
            builder.Property(p => p.IsActive).HasDefaultValue(true);
        }
    }
```

```csharp
    public class LeaderboardModel : IDbEntity<LeaderboardModel>, IDbActive
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Key]
        public string Id { get; set; }

        public string Name { get; set; }
        public virtual ICollection<UserLeaderboard> UserLeaderboards { get; set; } = new List<UserLeaderboard>();
        
        public bool? IsActive { get; set; }

        public virtual ICollection<ScoreModel> Scores { get; set; }

        public void OnModelCreating(EntityTypeBuilder<LeaderboardModel> builder)
        {
            // ensuring Name is unique
            builder.HasIndex(p => p.Name).IsUnique();
            builder.Property(p => p.IsActive).HasDefaultValue(true);
        }
    }
```

And this relationship table:

```csharp
    public class UserLeaderboard : IDbEntity<UserLeaderboard>
    {
        public string UserId { get; set; }
        public virtual UserProfileModel User { get; set; }

        public string LeaderboardId { get; set; }

        public virtual LeaderboardModel Leaderboard { get; set; }

        public void OnModelCreating(EntityTypeBuilder<UserLeaderboard> builder)
        {
            // User -> Leaderboard many-to-many
            builder.HasOne(pt => pt.User)
                .WithMany(p => p.UserLeaderboards)
                .HasForeignKey(pt => pt.UserId);

            // Leaderboard -> User many-to-many
            builder.HasOne(pt => pt.Leaderboard)
                .WithMany(t => t.UserLeaderboards)
                .HasForeignKey(pt => pt.LeaderboardId);

            
            builder.Property(u => u.UserId).ValueGeneratedNever().IsRequired();
            builder.Property(u => u.LeaderboardId).ValueGeneratedNever().IsRequired();

            // composite key
            builder.HasKey(e => new
            {
                e.UserId,
                e.LeaderboardId
            });
        }
    }
```

Then in my tests, I do the following:

```csharp

            // should throw duplicate key exception in the database
            profile.UserLeaderboards.Add(new UserLeaderboard {
                Leaderboard = leaderboards[0],
                User = profile
            });

            await Assert.ThrowsAsync<InvalidOperationException>(async () => await ctx.SaveChangesAsync());

            var board = await ctx.UserLeaderboards.AddAsync(new UserLeaderboard {
                Leaderboard = null,
                User = profile
            });

            await Assert.ThrowsAsync<InvalidOperationException>(async () => await ctx.SaveChangesAsync());
```

However, the second test doesn't throw a `InvalidOperationException`, but instead throws `DbUpdateException`. Even though Leaderboard is explicility set to null, it generates the LeaderboardId anyway when `AddAsync` is called. This is before the database it hit, so it must be that the framework is generating the keys even though I stated that neight should be auto generated. My guess is that having an actual key, and then indexing those two columns would work as expected, but I still consider this to be an issue.

### Further technical details

EF Core version: 3.1.1
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL
Operating system: Windows Host, mcr.microsoft.com/dotnet/core/sdk:3.1-bionic docker container
IDE: VSCode

**Comment by thomasrea0113** [comment]
```csharp


            profile.UserLeaderboards.Add(new UserLeaderboard {
                Leaderboard = leaderboards[0],
                User = profile
            });


            await manager.SaveChangesAsync();

            await ctx.Entry(profile).ReloadAsync();
            Assert.Contains(profile.UserLeaderboards, ul => ul.LeaderboardId == leaderboards[0].Id);

            // should throw duplicate key exception in the database
            profile.UserLeaderboards.Add(new UserLeaderboard {
                Leaderboard = leaderboards[0],
                User = profile
            });
            await Assert.ThrowsAsync<DbUpdateException>(async () => await ctx.SaveChangesAsync());

            Assert.Equal(2, profile.UserLeaderboards.Count);

            await ctx.Entry(profile).ReloadAsync();

            Assert.Single(profile.UserLeaderboards);
```

maybe I'm grossly misunderstanding how this works... but if a `DbUpdateException` is thrown due to a failed unqiue key constraint, and I reload the data, I see that the second item in the list was still added:

![image](https://user-images.githubusercontent.com/14018447/73791231-e8bf1400-476f-11ea-8a38-82259bf1b419.png)

however, the database indicates that only one exists!

![image](https://user-images.githubusercontent.com/14018447/73791295-0ab89680-4770-11ea-8112-4b1c24db9995.png)


what am I missing here?

**Comment by ajcvickers** [comment]
@thomasrea0113 It's not immediately clear to me what is going on and the code above is not complete enough for me to be able to reproduce what you are seeing. Please attach a small, runnable project or post a runnable code listing so that I can investigate.

**Comment by thomasrea0113** [comment]
> @thomasrea0113 It's not immediately clear to me what is going on and the code above is not complete enough for me to be able to reproduce what you are seeing. Please attach a small, runnable project or post a runnable code listing so that I can investigate.

Thanks for the quick response. I'm actually tracking my current project in a public repository. With VSCode and the docker extension, you can open it right up in the IDE and it'll be preconfigured to run the tests.

https://github.com/thomasrea0113/pr-leaderboard

The tests that are failing due to the mentioned issue are [here](https://github.com/thomasrea0113/pr-leaderboard/blob/master/Leaderboard.Tests/Models/RelatedTags.Tests.cs) and [here](https://github.com/thomasrea0113/pr-leaderboard/blob/master/Leaderboard.Tests/Models/UserProfileModel.Tests.cs#L137)

I can extract the important parts into a smaller project if necessary. It's still in the early stages, however, so maybe this is easy enough to follow along with. Thanks!

**Comment by thomasrea0113** [comment]
I went ahead and grabbed a screenshot that exactly demonstrates my issue. The second insert should (and does... kinda) fail because it'll have the same composite key.

![image](https://user-images.githubusercontent.com/14018447/73805160-e969a180-4793-11ea-8899-c760e39bb960.png)

from what I can gather, `AddAsync` goes ahead and inserts the second relationship into the list without validating against the key being unique. Once I commit to the database, it fails as expected, but the tracked collection is not immediately updated, so it still shows both as having been added (I set a breakpoint before the call to `SaveChangesAsync` and confirmed this). Adding a call to `await ctx.Entry(profile).ReloadAsync()` before the assert doesn't solve the issue. It still shows 2 `UserLeaderboard`s, even though only one exists in the database.

After some trial and error, it seems I have to do the follow to detach the entity and totally refresh it from the database:

```csharp
            ctx.Entry(profile).State = EntityState.Detached;
            profile = await ctx.UserProfiles.FindAsync(profile.UserId);
```

So the only way to sync back up is to query the database AGAIN? It seems to be that the framework should see the exception get thrown and know to update the tracked entities

**Comment by thomasrea0113** [comment]
I'm actually going to go ahead and close this. I realized I was more or less misusing the DbContext. The DbContext is injected as a scoped dependency for a reason - it's not really meant to persist for more than one call to `SaveChangesAsync`. If you're database update fails, then in most scenarios you're termiante the request and let the user know. There's no real reason for the locally tracked objects to be updated if the save fails. I restructured my tests to be more inline with this use case, and everything now works as expected.
